### PR TITLE
Adding the NorthSouth API for Public Preview

### PR DESCRIPTION
**Pull request by MatanShabtay**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
@MatanShabtay have these APIs been reviewed by ARM?

**Comment by MatanShabtay** [comment]
> @MatanShabtay have these APIs been reviewed by ARM?

This resource type exists in the manifest and it's swagger was reviewed in the PR below, however the PR wasn't completed as this resource is currently in private preview.

We are working towards public preview and following recent changes in the requirements, the API completely changed.
Note that its not considered as a breaking change since we are still under private preview.

https://github.com/Azure/azure-rest-api-specs/pull/3669 

Once you confirm the PR please ask ARM to review as well.
Thanks.

**Comment by jhendrixMSFT** [comment]
@ravbhatnagar FYI for ARM review.

**Comment by MatanShabtay** [comment]
Hi,

Apologies for the late push but I just noticed that I forgot to add a POST action on the resource so I've added it.
Hopefully I didn't cause any trouble.

Thanks in advance.

**Comment by KrisBash** [comment]
@MatanShabtay - let's schedule a quick meeting to understand these changes (ARM api review)

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json
Comment (KrisBash) [inline specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json]: should be an apostrophe rather than semi-colon

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json
Comment (KrisBash) [inline specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json]: does 'allow**ed**RemoteAddresses' make more sense than 'allowRemoteAddresses'

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json
Comment (KrisBash) [inline specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json]: For these GET apis, I would assume that the response could be quite large, as each instance includes a number of rules and a number of alerts. It may be prudent to return only the most relevant data on GET, unless a querystring parameter ($expand) is passed. Additionally, supporting a $filter property (described [here](https://github.com/Microsoft/api-guidelines/blob/vNext/Guidelines.md#97-filtering)) can be helpful to reduce the response size.

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json
Comment (KrisBash) [inline specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json]: How many alerts can there be? Is there any risk of exceeding page size?

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json
Comment (KrisBash) [inline specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json]: This is an async operation, but I don't see the [x-ms-long-running-operation ](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-long-running-operation) extension

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json
Comment (KrisBash) [inline specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json]: I sent a question regarding authorization for this operation offline.

**Review by KrisBash — CHANGES_REQUESTED** [review]
Please review my comments

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json
Comment (MatanShabtay) [reply specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/security.json]: allowed (past tense) hints that the trafficHardeningRules already applies on the VM's NSG while these set of rules hasn't applied yet on the NSG (some/all of the rules may be enforced on the NSG using the Enforce action).

**Comment by jhendrixMSFT** [comment]
@MatanShabtay any update?

**Comment by MatanShabtay** [comment]
Hey,

We have a planned meeting with @KrisBash about the authorization.
We will update the PR and move on after we will make some decisions together.

**Comment by MatanShabtay** [comment]
1.Due to marketing descisions, the resource type name changed from 'northSouthHardenings' to 'adaptiveNetworkControls'
2.Following further discussions with ARM team - we understood that a resource routingType cannot be both 'Fanout' and 'Extension', so we've decided to stick with extension and removed all subscription and resource group scoped requests.
3.Following our offline discussion with @KrisBash , we have added the NSG IDs to the 'enforce' action request body so we will perform linkedAccessCheck in order to respect RBAC authorizations.
4.Removed the 'alerts' property from the resource - we plan to consume the alerts from a different flow in Microsoft.Security (alerts API)

**Review by KrisBash — APPROVED** [review]
LGTM - thanks!

**Comment by MatanShabtay** [comment]
> @MatanShabtay looks good from SDK perspective. Has this functionality been deployed yet?

We are working on updating Microsoft.Security manifest.
Will report back in approx 2weeks (still under development)

**Comment by jhendrixMSFT** [comment]
@MatanShabtay any update?  Also there's a merge conflict that needs to be resolved.

**Comment by MatanShabtay** [comment]
> @MatanShabtay any update? Also there's a merge conflict that needs to be resolved.

Hi,

Due to marketing decisions the respirce name is about to change from adaptiveNetworkControls to adaptiveNetworkHardenings.

We will soon update the PR with the new resource name.
Should there be any concern about us changing the name, please let me know.

Thanks.

**Comment by MatanShabtay** [comment]
Hi @jhendrixMSFT & @KrisBash , Due to marketing decisions I have renamed the resource type name to adaptiveNetworkHardenings from adaptiveNetworkControls & resolved the merge conflicts.

Please review before we will update the manifest and add the new resource type

**Review by jhendrixMSFT — CHANGES_REQUESTED** [review]

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json
Comment (MatanShabtay) [reply specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json]: Thanks!
I've renamed the paths and parameters so only collection operation and objects will be named in plural

File: specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json
Comment (jhendrixMSFT) [inline specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json]: Should be `AdaptiveNetworkHardenings_Get` (missing 's').

Comment (MatanShabtay) [reply specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json]: Why? it gets only a single resource, not a list

Comment (jhendrixMSFT) [reply specification/security/resource-manager/Microsoft.Security/preview/2015-06-01-preview/adaptiveNetworkHardenings.json]: For a given operation ID, AutoRest generates an operation group (e.g. an interface in C#) that's named to everything left of the '_' character that will contain all of the methods in the same group.  So if you have the following operation groups.
```
Foos_Get
Foos_CreateOrUpdate
Foos_Delete
```
You'd get the following codegen.
```C#
public interface Foo {
    Get();
    CreateOrUpdate();
    Delete();
}
```
The other problem here is that operation group name `AdaptiveNetworkHardening` collides with the type `AdaptiveNetworkHardening` defined on line 324; this will force the code generators to disambiguate the name by whatever means the generators have implemented (usually some generic prefix or suffix) which is not very optimal.  It's for this reason that the authoring guidelines state that for operation groups NOUN_VERB the noun be plural so the model type can be singular.

**Comment by MatanShabtay** [comment]
Hey @jhendrixMSFT , waiting for your approval.
Can you then refer someone from ARV to review?

Thanks.

**Review by jhendrixMSFT — APPROVED** [review]

**Comment by jhendrixMSFT** [comment]
@MatanShabtay has this functionality been deployed yet?  I still see DoNotMerge label

**Review by KrisBash — APPROVED** [review]
Thanks for working with us on this review. LGTM

**Comment by adxsdknet** [comment]
# Automation for azure-sdk-for-net 
A PR has been created for you: 
https://github.com/azure/azure-sdk-for-net/pull/5339 
.NET SDK Commits: 
https://github.com/adxsdknet/azure-sdk-for-net/pull/5339/commits/3f8ddd476ad09fbab1ef163efae0a0a36dd6a2bb

**Comment by jhendrixMSFT** [comment]
@MatanShabtay hello any update?

**Comment by MatanShabtay** [comment]
Hi @jhendrixMSFT ,

The API has been deployed and we ready to complete this PR.
Thanks.

**Comment by jhendrixMSFT** [comment]
@MatanShabtay I don't see the new swagger referenced in the readme.md file so some SDK won't get the new APIs generated for them.  Is this intentional?

**Comment by MatanShabtay** [comment]
> @MatanShabtay I don't see the new swagger referenced in the readme.md file so some SDK won't get the new APIs generated for them. Is this intentional?

Hey, I've fixed it.
I see that the Avocado check is failing but based on the error message it doesnt to be related to my PR.

Can you please confirm this ?

**Comment by jhendrixMSFT** [comment]
@MatanShabtay it's not but has been fixed in a separate PR.

**Pull request merged**
